File: SRC.NET/ListBox.vb
Option Strict Off
Option Explicit On
Friend Class frmListBox
	Inherits System.Windows.Forms.Form
	
	' Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
	' 本プログラムはフリーソフトであり、無保証です。
	' 本プログラムはGNU General Public License(Ver.3またはそれ以降)が定める条件の下で
	' 再頒布または改変することができます。
	
	'リストボックスのフォーム
	
	'リストボックスのサイズ (通常:M 大型:L)
	'幅
	Public HorizontalSize As String
	'高さ
	Public VerticalSize As String
	
	'Questionコマンド用変数
	Public CurrentTime As Short
	Public TimeLimit As Short
	
	'最後に選択されたアイテム
	Private LastSelectedItem As Short
	
	'リストボックスへのキー入力
	Private Sub frmListBox_KeyDown(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
		Dim KeyCode As Short = eventArgs.KeyCode
		Dim Shift As Short = eventArgs.KeyData \ &H10000
		'既にウィンドウが隠れている場合は無視
		If Not Visible Then
			Exit Sub
		End If
		
		Select Case KeyCode
			Case System.Windows.Forms.Keys.Up
			Case System.Windows.Forms.Keys.Down
			Case System.Windows.Forms.Keys.Left, System.Windows.Forms.Keys.Right
			Case System.Windows.Forms.Keys.Escape, System.Windows.Forms.Keys.Delete, System.Windows.Forms.Keys.Back
				'キャンセル
				SelectedItem = 0
				LastSelectedItem = SelectedItem
				TopItem = lstItems.TopIndex
				If IsFormClicked Then
					Hide()
				End If
				IsFormClicked = True
			Case Else
				'選択
				If lstItems.SelectedIndex < 0 Then
					Exit Sub
				End If
				If UBound(ListItemFlag) >= lstItems.SelectedIndex + 1 Then
					If ListItemFlag(lstItems.SelectedIndex + 1) Then
						Exit Sub
					End If
				End If
				SelectedItem = lstItems.SelectedIndex + 1
				LastSelectedItem = SelectedItem
				TopItem = lstItems.TopIndex + 1
				If IsFormClicked Then
					Hide()
				End If
				IsFormClicked = True
		End Select
	End Sub
	
	'リストボックスを開く
	Private Sub frmListBox_Load(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles MyBase.Load
		Dim ret As Integer
		
		'リストボックスを常に手前に表示
		ret = SetWindowPos(Handle.ToInt32, -1, 0, 0, 0, 0, &H3)
		HorizontalSize = "M"
		VerticalSize = "M"
	End Sub
	
	'リストボックスを閉じる
	Private Sub frmListBox_FormClosed(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.FormClosedEventArgs) Handles Me.FormClosed
		TopItem = lstItems.TopIndex + 1
		IsFormClicked = True
		If Not IsMordal And Visible Then
			'UPGRADE_ISSUE: Event パラメータ Cancel はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="FB723E3C-1C06-4D2B-B083-E6CD0D334DA8"' をクリックしてください。
			Cancel = 1
		End If
		Hide()
	End Sub
	
	'項目をダブルクリック
	Private Sub lstItems_DoubleClick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles lstItems.DoubleClick
		'無効なアイテムが選択されている？
		If lstItems.SelectedIndex < 0 Then
			Exit Sub
		End If
		If UBound(ListItemFlag) >= lstItems.SelectedIndex + 1 Then
			If ListItemFlag(lstItems.SelectedIndex + 1) Then
				Exit Sub
			End If
		End If
		
		If LastSelectedItem <> 0 Then
			'連続で選択
			If Not Visible Then
				Exit Sub
			End If
			SelectedItem = lstItems.SelectedIndex + 1
			LastSelectedItem = SelectedItem
			TopItem = lstItems.TopIndex + 1
			If IsFormClicked Then
				Hide()
			End If
			IsFormClicked = True
		Else
			'連続でキャンセル
			SelectedItem = 0
			LastSelectedItem = SelectedItem
			TopItem = lstItems.TopIndex + 1
			If IsFormClicked Then
				Hide()
			End If
			IsFormClicked = True
		End If
	End Sub
	
	'マウスでクリック
	Private Sub lstItems_MouseDown(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.MouseEventArgs) Handles lstItems.MouseDown
		Dim Button As Short = eventArgs.Button \ &H100000
		Dim Shift As Short = System.Windows.Forms.Control.ModifierKeys \ &H10000
		Dim X As Single = VB6.PixelsToTwipsX(eventArgs.X)
		Dim Y As Single = VB6.PixelsToTwipsY(eventArgs.Y)
		Select Case Button
			Case 1
				'選択
				If Not Visible Then
					Exit Sub
				End If
				
				'無効なアイテムが選択されている？
				If lstItems.SelectedIndex < 0 Then
					Exit Sub
				End If
				If UBound(ListItemFlag) >= lstItems.SelectedIndex + 1 Then
					If ListItemFlag(lstItems.SelectedIndex + 1) Then
						Exit Sub
					End If
				End If
				
				SelectedItem = lstItems.SelectedIndex + 1
				LastSelectedItem = SelectedItem
				TopItem = lstItems.TopIndex + 1
				IsFormClicked = True
				
				If CurrentTime > 0 Then
					Timer1.Enabled = False
					Hide()
				End If
				
			Case 2
				'キャンセル
				SelectedItem = 0
				LastSelectedItem = SelectedItem
				TopItem = lstItems.TopIndex + 1
				IsFormClicked = True
		End Select
	End Sub
	
	'キャプション部分をクリック
	Private Sub labCaption_MouseDown(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.MouseEventArgs) Handles labCaption.MouseDown
		Dim Button As Short = eventArgs.Button \ &H100000
		Dim Shift As Short = System.Windows.Forms.Control.ModifierKeys \ &H10000
		Dim X As Single = VB6.PixelsToTwipsX(eventArgs.X)
		Dim Y As Single = VB6.PixelsToTwipsY(eventArgs.Y)
		Select Case Button
			Case 1
				'ユニットステータスを表示しているユニットを入れ替え
				If MainForm.Visible Then
					If Not DisplayedUnit Is Nothing And Not SelectedUnit Is Nothing And Not SelectedTarget Is Nothing Then
						If DisplayedUnit.ID = SelectedUnit.ID Then
							DisplayUnitStatus(SelectedTarget)
						Else
							DisplayUnitStatus(SelectedUnit)
						End If
					End If
				End If
			Case 2
				'キャンセル
				SelectedItem = 0
				LastSelectedItem = SelectedItem
				TopItem = lstItems.TopIndex
				If IsFormClicked Then
					Hide()
				End If
				IsFormClicked = True
		End Select
	End Sub
	
	'リストボックスの端をクリック
	Private Sub frmListBox_MouseDown(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.MouseEventArgs) Handles MyBase.MouseDown
		Dim Button As Short = eventArgs.Button \ &H100000
		Dim Shift As Short = System.Windows.Forms.Control.ModifierKeys \ &H10000
		Dim X As Single = eventArgs.X
		Dim Y As Single = eventArgs.Y
		If Button = 2 Then
			'キャンセル
			SelectedItem = 0
			LastSelectedItem = SelectedItem
			TopItem = lstItems.TopIndex
			If IsFormClicked Then
				Hide()
			End If
			IsFormClicked = True
		End If
	End Sub
	
	'リストボックス上でマウスカーソルを動かす
	Private Sub lstItems_MouseMove(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.MouseEventArgs) Handles lstItems.MouseMove
		Dim Button As Short = eventArgs.Button \ &H100000
		Dim Shift As Short = System.Windows.Forms.Control.ModifierKeys \ &H10000
		Dim X As Single = VB6.PixelsToTwipsX(eventArgs.X)
		Dim Y As Single = VB6.PixelsToTwipsY(eventArgs.Y)
		Dim itm As Short
		
		'カーソルがあるアイテムを算出
		itm = ((Y * ClientRectangle.Width) \ VB6.PixelsToTwipsX(Width) + 1) \ 16
		itm = itm + lstItems.TopIndex
		
		'カーソル上のアイテムをハイライト表示
		If itm < 0 Or itm >= lstItems.Items.Count Then
			lstItems.SelectedIndex = -1
			Exit Sub
		End If
		If lstItems.SelectedIndex = itm Then
			Exit Sub
		End If
		lstItems.SelectedIndex = itm
		
		'コメント欄を更新
		If txtComment.Enabled Then
			txtComment.Text = ListItemComment(itm + 1)
		End If
	End Sub
	
	'Questionコマンド対応
	Private Sub Timer1_Tick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Timer1.Tick
		CurrentTime = CurrentTime + 1
		'UPGRADE_ISSUE: PictureBox メソッド picBar.Cls はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="CC4C7EC0-C903-48FC-ACCC-81861D12DA4A"' をクリックしてください。
		picBar.Cls()
		'UPGRADE_ISSUE: PictureBox メソッド picBar.Line はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="CC4C7EC0-C903-48FC-ACCC-81861D12DA4A"' をクリックしてください。
		picBar.Line (0, 0) - (picBar.ClientRectangle.Width * CurrentTime \ TimeLimit, picBar.ClientRectangle.Height), BF
		picBar.Refresh()
		
		If CurrentTime >= TimeLimit Then
			SelectedItem = 0
			LastSelectedItem = SelectedItem
			TopItem = lstItems.TopIndex
			Hide()
			Timer1.Enabled = False
		End If
	End Sub
	
	'選択されているアイテムに対応するユニットのステータス表示
	Private Sub Timer2_Tick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Timer2.Tick
		Dim u As Unit
		
		If Not Visible Or Not MainForm.Visible Then
			Exit Sub
		End If
		
		If lstItems.SelectedIndex = -1 Then
			Exit Sub
		End If
		
		On Error GoTo ErrorHandler
		
		If lstItems.SelectedIndex >= UBound(ListItemID) Then
			Exit Sub
		End If
		
		'選択されたユニットが存在する？
		If Not UList.IsDefined2(ListItemID(lstItems.SelectedIndex + 1)) Then
			Exit Sub
		End If
		
		u = UList.Item2(ListItemID(lstItems.SelectedIndex + 1))
		
		'選択されたユニットにパイロットが乗っている？
		If u.CountPilot = 0 Then
			Exit Sub
		End If
		
		'既に表示している？
		If DisplayedUnit Is u Then
			Exit Sub
		End If
		
		'選択されたユニットをステータスウィンドウに表示
		DisplayUnitStatus(u)
		
ErrorHandler: 
	End Sub
End Class

File: SRC.Sharp/SRCTestForm/Forms/ListBox.cs
using System;
using System.Windows.Forms;
using Microsoft.VisualBasic;

namespace Project1
{
    internal partial class frmListBox : Form
    {

        // Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
        // 本プログラムはフリーソフトであり、無保証です。
        // 本プログラムはGNU General Public License(Ver.3またはそれ以降)が定める条件の下で
        // 再頒布または改変することができます。

        // リストボックスのフォーム

        // リストボックスのサイズ (通常:M 大型:L)
        // 幅
        public string HorizontalSize;
        // 高さ
        public string VerticalSize;

        // Questionコマンド用変数
        public short CurrentTime;
        public short TimeLimit;

        // 最後に選択されたアイテム
        private short LastSelectedItem;

        // リストボックスへのキー入力
        private void frmListBox_KeyDown(object eventSender, KeyEventArgs eventArgs)
        {
            short KeyCode = (short)eventArgs.KeyCode;
            short Shift = (short)((int)eventArgs.KeyData / 0x10000);
            // 既にウィンドウが隠れている場合は無視
            if (!Visible)
            {
                return;
            }

            switch (KeyCode)
            {
                case (short)Keys.Up:
                    {
                        break;
                    }

                case (short)Keys.Down:
                    {
                        break;
                    }

                case (short)Keys.Left:
                case (short)Keys.Right:
                    {
                        break;
                    }

                case (short)Keys.Escape:
                case (short)Keys.Delete:
                case (short)Keys.Back:
                    {
                        // キャンセル
                        Commands.SelectedItem = 0;
                        LastSelectedItem = Commands.SelectedItem;
                        GUI.TopItem = (short)lstItems.TopIndex;
                        if (GUI.IsFormClicked)
                        {
                            Hide();
                        }

                        GUI.IsFormClicked = true;
                        break;
                    }

                default:
                    {
                        // 選択
                        if (lstItems.SelectedIndex < 0)
                        {
                            return;
                        }

                        if (Information.UBound(GUI.ListItemFlag) >= lstItems.SelectedIndex + 1)
                        {
                            if (GUI.ListItemFlag[lstItems.SelectedIndex + 1])
                            {
                                return;
                            }
                        }

                        Commands.SelectedItem = (short)(lstItems.SelectedIndex + 1);
                        LastSelectedItem = Commands.SelectedItem;
                        GUI.TopItem = (short)(lstItems.TopIndex + 1);
                        if (GUI.IsFormClicked)
                        {
                            Hide();
                        }

                        GUI.IsFormClicked = true;
                        break;
                    }
            }
        }

        // リストボックスを開く
        private void frmListBox_Load(object eventSender, EventArgs eventArgs)
        {
            int ret;

            // リストボックスを常に手前に表示
            ret = GUI.SetWindowPos(Handle.ToInt32(), -1, 0, 0, 0, 0, 0x3);
            HorizontalSize = "M";
            VerticalSize = "M";
        }

        // リストボックスを閉じる
        private void frmListBox_FormClosed(object eventSender, FormClosedEventArgs eventArgs)
        {
            GUI.TopItem = (short)(lstItems.TopIndex + 1);
            GUI.IsFormClicked = true;
            if (!GUI.IsMordal & Visible)
            {
                // UPGRADE_ISSUE: Event パラメータ Cancel はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="FB723E3C-1C06-4D2B-B083-E6CD0D334DA8"' をクリックしてください。
                Cancel = 1;
            }

            Hide();
        }

        // 項目をダブルクリック
        private void lstItems_DoubleClick(object eventSender, EventArgs eventArgs)
        {
            // 無効なアイテムが選択されている？
            if (lstItems.SelectedIndex < 0)
            {
                return;
            }

            if (Information.UBound(GUI.ListItemFlag) >= lstItems.SelectedIndex + 1)
            {
                if (GUI.ListItemFlag[lstItems.SelectedIndex + 1])
                {
                    return;
                }
            }

            if (LastSelectedItem != 0)
            {
                // 連続で選択
                if (!Visible)
                {
                    return;
                }

                Commands.SelectedItem = (short)(lstItems.SelectedIndex + 1);
                LastSelectedItem = Commands.SelectedItem;
                GUI.TopItem = (short)(lstItems.TopIndex + 1);
                if (GUI.IsFormClicked)
                {
                    Hide();
                }

                GUI.IsFormClicked = true;
            }
            else
            {
                // 連続でキャンセル
                Commands.SelectedItem = 0;
                LastSelectedItem = Commands.SelectedItem;
                GUI.TopItem = (short)(lstItems.TopIndex + 1);
                if (GUI.IsFormClicked)
                {
                    Hide();
                }

                GUI.IsFormClicked = true;
            }
        }

        // マウスでクリック
        private void lstItems_MouseDown(object eventSender, MouseEventArgs eventArgs)
        {
            short Button = (short)((int)eventArgs.Button / 0x100000);
            short Shift = (short)((int)ModifierKeys / 0x10000);
            float X = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsX(eventArgs.X);
            float Y = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsY(eventArgs.Y);
            switch (Button)
            {
                case 1:
                    {
                        // 選択
                        if (!Visible)
                        {
                            return;
                        }

                        // 無効なアイテムが選択されている？
                        if (lstItems.SelectedIndex < 0)
                        {
                            return;
                        }

                        if (Information.UBound(GUI.ListItemFlag) >= lstItems.SelectedIndex + 1)
                        {
                            if (GUI.ListItemFlag[lstItems.SelectedIndex + 1])
                            {
                                return;
                            }
                        }

                        Commands.SelectedItem = (short)(lstItems.SelectedIndex + 1);
                        LastSelectedItem = Commands.SelectedItem;
                        GUI.TopItem = (short)(lstItems.TopIndex + 1);
                        GUI.IsFormClicked = true;
                        if (CurrentTime > 0)
                        {
                            Timer1.Enabled = false;
                            Hide();
                        }

                        break;
                    }

                case 2:
                    {
                        // キャンセル
                        Commands.SelectedItem = 0;
                        LastSelectedItem = Commands.SelectedItem;
                        GUI.TopItem = (short)(lstItems.TopIndex + 1);
                        GUI.IsFormClicked = true;
                        break;
                    }
            }
        }

        // キャプション部分をクリック
        private void labCaption_MouseDown(object eventSender, MouseEventArgs eventArgs)
        {
            short Button = (short)((int)eventArgs.Button / 0x100000);
            short Shift = (short)((int)ModifierKeys / 0x10000);
            float X = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsX(eventArgs.X);
            float Y = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsY(eventArgs.Y);
            switch (Button)
            {
                case 1:
                    {
                        // ユニットステータスを表示しているユニットを入れ替え
                        if (GUI.MainForm.Visible)
                        {
                            if (Status.DisplayedUnit is object & Commands.SelectedUnit is object & Commands.SelectedTarget is object)
                            {
                                if ((Status.DisplayedUnit.ID ?? "") == (Commands.SelectedUnit.ID ?? ""))
                                {
                                    Status.DisplayUnitStatus(ref Commands.SelectedTarget);
                                }
                                else
                                {
                                    Status.DisplayUnitStatus(ref Commands.SelectedUnit);
                                }
                            }
                        }

                        break;
                    }

                case 2:
                    {
                        // キャンセル
                        Commands.SelectedItem = 0;
                        LastSelectedItem = Commands.SelectedItem;
                        GUI.TopItem = (short)lstItems.TopIndex;
                        if (GUI.IsFormClicked)
                        {
                            Hide();
                        }

                        GUI.IsFormClicked = true;
                        break;
                    }
            }
        }

        // リストボックスの端をクリック
        private void frmListBox_MouseDown(object eventSender, MouseEventArgs eventArgs)
        {
            short Button = (short)((int)eventArgs.Button / 0x100000);
            short Shift = (short)((int)ModifierKeys / 0x10000);
            float X = eventArgs.X;
            float Y = eventArgs.Y;
            if (Button == 2)
            {
                // キャンセル
                Commands.SelectedItem = 0;
                LastSelectedItem = Commands.SelectedItem;
                GUI.TopItem = (short)lstItems.TopIndex;
                if (GUI.IsFormClicked)
                {
                    Hide();
                }

                GUI.IsFormClicked = true;
            }
        }

        // リストボックス上でマウスカーソルを動かす
        private void lstItems_MouseMove(object eventSender, MouseEventArgs eventArgs)
        {
            short Button = (short)((int)eventArgs.Button / 0x100000);
            short Shift = (short)((int)ModifierKeys / 0x10000);
            float X = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsX(eventArgs.X);
            float Y = (float)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsY(eventArgs.Y);
            short itm;

            // カーソルがあるアイテムを算出
            itm = (short)(((long)(Y * ClientRectangle.Width) / (long)Microsoft.VisualBasic.Compatibility.VB6.Support.PixelsToTwipsX(Width) + 1L) / 16L);
            itm = (short)(itm + lstItems.TopIndex);

            // カーソル上のアイテムをハイライト表示
            if (itm < 0 | itm >= lstItems.Items.Count)
            {
                lstItems.SelectedIndex = -1;
                return;
            }

            if (lstItems.SelectedIndex == itm)
            {
                return;
            }

            lstItems.SelectedIndex = itm;

            // コメント欄を更新
            if (txtComment.Enabled)
            {
                txtComment.Text = GUI.ListItemComment[itm + 1];
            }
        }

        // Questionコマンド対応
        private void Timer1_Tick(object eventSender, EventArgs eventArgs)
        {
            CurrentTime = (short)(CurrentTime + 1);
            // UPGRADE_ISSUE: PictureBox メソッド picBar.Cls はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="CC4C7EC0-C903-48FC-ACCC-81861D12DA4A"' をクリックしてください。
            picBar.Cls();
            // UPGRADE_ISSUE: PictureBox メソッド picBar.Line はアップグレードされませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="CC4C7EC0-C903-48FC-ACCC-81861D12DA4A"' をクリックしてください。
            picBar.Line(0, 0); /* TODO ERROR: Skipped SkippedTokensTrivia *//* TODO ERROR: Skipped SkippedTokensTrivia */
            picBar.Refresh();
            if (CurrentTime >= TimeLimit)
            {
                Commands.SelectedItem = 0;
                LastSelectedItem = Commands.SelectedItem;
                GUI.TopItem = (short)lstItems.TopIndex;
                Hide();
                Timer1.Enabled = false;
            }
        }

        // 選択されているアイテムに対応するユニットのステータス表示
        private void Timer2_Tick(object eventSender, EventArgs eventArgs)
        {
            Unit u;
            if (!Visible | !GUI.MainForm.Visible)
            {
                return;
            }

            if (lstItems.SelectedIndex == -1)
            {
                return;
            };
#error Cannot convert OnErrorGoToStatementSyntax - see comment for details
            /* Cannot convert OnErrorGoToStatementSyntax, CONVERSION ERROR: Conversion for OnErrorGoToLabelStatement not implemented, please report this issue in 'On Error GoTo ErrorHandler' at character 10644


            Input:

                    On Error GoTo ErrorHandler

             */
            if (lstItems.SelectedIndex >= Information.UBound(GUI.ListItemID))
            {
                return;
            }

            // 選択されたユニットが存在する？
            bool localIsDefined2() { var tmp = GUI.ListItemID; object argIndex1 = tmp[lstItems.SelectedIndex + 1]; var ret = SRC.UList.IsDefined2(ref argIndex1); return ret; }

            if (!localIsDefined2())
            {
                return;
            }

            var tmp = GUI.ListItemID;
            object argIndex1 = tmp[lstItems.SelectedIndex + 1];
            u = SRC.UList.Item2(ref argIndex1);

            // 選択されたユニットにパイロットが乗っている？
            if (u.CountPilot() == 0)
            {
                return;
            }

            // 既に表示している？
            if (ReferenceEquals(Status.DisplayedUnit, u))
            {
                return;
            }

            // 選択されたユニットをステータスウィンドウに表示
            Status.DisplayUnitStatus(ref u);
            ErrorHandler:
            ;
        }
    }
}